ledger/test/test_compact_serializer.py

```diff
@@ -0,0 +1,228 @@
+
+
+def testDeserializeLessFields():
+    serializer = CompactSerializer(fields)
+
+    assert {"f1": None, "f2": "v1", "f3": 2, "f4": None, "f5": None} == serializer.deserialize(b"|v1|2||")
+    assert {"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": None} == serializer.deserialize(b"v1||||")
+    assert {"f1": None, "f2": None, "f3": None, "f4": None, "f5": None} == serializer.deserialize(b"|||||")
+    assert {"f1": None, "f2": None, "f3": None, "f4": None, "f5": True} == serializer.deserialize(b"||||True" )
+    assert {"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": False} == serializer.deserialize(b"v1||||False")
+
+
+def testSerializeLessFieldsWithNone():
+    serializer = CompactSerializer(fields)
+
+    assert b"|v1|2||" == serializer.serialize({"f1": None, "f2": "v1", "f3": 2})
+    assert b"v1||||" == serializer.serialize({"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": None})
+    assert b"||||" == serializer.serialize({"f1": None, "f2": None, "f3": None, "f4": None, "f5": None})
+    assert b"||||3" == serializer.serialize({"f1": None, "f5": 3})
+    assert b"||||3" == serializer.serialize({"f1": None, "f2": None, "f3": None, "f4": None, "f5": 3})
+    assert b"v1||||3" == serializer.serialize({"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": 3})
+    assert b"v1||||3" == serializer.serialize({"f1": "v1", "f3": None, "f5": 3})
+
+def testSerializeInAnyOrder():
+    serializer = CompactSerializer(fields)
+
+    assert b"|v1|2||" == serializer.serialize({"f3": 2, "f2": "v1"})
+    assert b"v1||||3" == serializer.serialize({"f5": 3, "f1": "v1", })
+    assert b"v1|v2|3|4.0|True" == \
+           serializer.serialize(
+               {"f2": "v2", "f5": True, "f3": 3, "f1": "v1", "f4": 4.0}
+           )
+
+def testSerializeSubfields():
+    fields = OrderedDict([
+        ("f1.a", (str, str)),
+        ("f1.b", (str, int)),
+        ("f1.c", (str, float)),
+        ("f2.d", (str, str)),
+        ("f2.e", (str, int)),
+        ("f2.f", (str, float)),
+
+    ])
+    serializer = CompactSerializer(fields)
+
+    json = {
+        "f1":{"a": "v1", "b": 2, "c": 3.0},
+        "f2": {"d": "v1", "e": 3, "f": 4.0},
+    }
+    assert b"v1|2|3.0|v1|3|4.0" == serializer.serialize(json)
+
+
+def testDeserializeSubfields():
+    fields = OrderedDict([
+        ("f1.a", (str, str)),
+        ("f1.b", (str, int)),
+        ("f1.c", (str, float)),
+        ("f2.d", (str, str)),
+        ("f2.e", (str, int)),
+        ("f2.f", (str, float)),
+
+    ])
+    serializer = CompactSerializer(fields)
+
+    json = {
+        "f1":{"a": "v1", "b": 2, "c": 3.0},
+        "f2": {"d": "v1", "e": 3, "f": 4.0},
+    }
+    assert json == serializer.deserialize(b"v1|2|3.0|v1|3|4.0")
+
+
+def testSerializeWrongFields():
+    serializer = CompactSerializer(fields)
+    assert b"||||" == serializer.serialize({"wrongField": "v1"})
+    assert b"||||" == serializer.serialize({"wrongField1": "v1", "wrongField2": "v2"})
```

Comment:
A wrong field always results in an empty string in the serialisation, and since serialisation of each field is independent, trying several permutations of these should not be necessary, anyway that's my opinion, the only problem i see with these is that it takes more time to complete tests